File: app/src/main/java/com/aero51/moviedatabase/utils/Converters.java
package com.aero51.moviedatabase.utils;

import androidx.room.TypeConverter;

import com.aero51.moviedatabase.repository.model.omdb.OmdbModel;
import com.aero51.moviedatabase.repository.model.tmdb.credits.ActorSearchResponse;
import com.aero51.moviedatabase.repository.model.tmdb.movie.MovieDetailsResponse;
import com.aero51.moviedatabase.repository.model.tmdb.movie.MovieGenresResponse;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowDetailsResponse;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowGenresResponse;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.Date;
import java.util.List;


//Room cannot store lists and Date among others  so type converters must be implemented
public class Converters {
    @TypeConverter
    public static Date fromTimestamp(Long value) {
        return value == null ? null : new Date(value);
    }

    @TypeConverter
    public static Long dateToTimestamp(Date date) {
        return date == null ? null : date.getTime();
    }


    @TypeConverter
    public static List<Integer> fromString(String value) {
        Type listType = new TypeToken<List<Integer>>() {
        }.getType();
        return new Gson().fromJson(value, listType);
    }

    @TypeConverter
    public static String fromArrayList(List<Integer> list) {
        Gson gson = new Gson();
        String json = gson.toJson(list);
        return json;
    }

    @TypeConverter
    public static List<String> fromString2(String value) {
        Type listType = new TypeToken<List<String>>() {
        }.getType();
        return new Gson().fromJson(value, listType);
    }

    @TypeConverter
    public static String fromList(List<String> list) {
        Gson gson = new Gson();
        String json = gson.toJson(list);
        return json;
    }


    @TypeConverter
    public static List<OmdbModel.Ratings> toRatingsList(String value) {
        if (value == null) {
            return (null);
        }
        Gson gson = new Gson();
        Type type = new TypeToken<List<OmdbModel.Ratings>>() {
        }.getType();
        List<OmdbModel.Ratings> ratingsList = gson.fromJson(value, type);
        return ratingsList;
    }

    @TypeConverter
    public static String fromRatingsList(List<OmdbModel.Ratings> ratings) {
        Gson gson = new Gson();
        String json = gson.toJson(ratings);
        return json;
    }

    @TypeConverter
    public static List<MovieDetailsResponse.ProductionCompany> toMovieProductionCompaniesList(String value) {
        if (value == null) {
            return (null);
        }
        Gson gson = new Gson();
        Type type = new TypeToken<List<MovieDetailsResponse.ProductionCompany>>() {
        }.getType();
        List<MovieDetailsResponse.ProductionCompany> productionCompanies = gson.fromJson(value, type);
        return productionCompanies;
    }

    @TypeConverter
    public static String fromMovieProductionCompaniesList(List<MovieDetailsResponse.ProductionCompany> productionCompanies) {
        Gson gson = new Gson();
        String json = gson.toJson(productionCompanies);
        return json;
    }

    @TypeConverter
    public static List<TvShowDetailsResponse.ProductionCompany> toTvShowProductionCompaniesList(String value) {
        if (value == null) {
            return (null);
        }
        Gson gson = new Gson();
        Type type = new TypeToken<List<TvShowDetailsResponse.ProductionCompany>>() {
        }.getType();
        List<TvShowDetailsResponse.ProductionCompany> productionCompanies = gson.fromJson(value, type);
        return productionCompanies;
    }

    @TypeConverter
    public static String fromTvShowProductionCompaniesList(List<TvShowDetailsResponse.ProductionCompany> productionCompanies) {
        Gson gson = new Gson();
        String json = gson.toJson(productionCompanies);
        return json;
    }

    @TypeConverter
    public static List<MovieGenresResponse.MovieGenre> toGenresForMovieList(String value) {
        if (value == null) {
            return (null);
        }
        Gson gson = new Gson();
        Type type = new TypeToken<List<MovieGenresResponse.MovieGenre>>() {
        }.getType();
        List<MovieGenresResponse.MovieGenre> movieGenres = gson.fromJson(value, type);
        return movieGenres;
    }

    @TypeConverter
    public static String fromGenreForMovieList(List<MovieGenresResponse.MovieGenre> movieGenres) {
        Gson gson = new Gson();
        String json = gson.toJson(movieGenres);
        return json;
    }

    @TypeConverter
    public static List<TvShowGenresResponse.TvShowGenre> toGenresForTvShowList(String value) {
        if (value == null) {
            return (null);
        }
        Gson gson = new Gson();
        Type type = new TypeToken<List<TvShowGenresResponse.TvShowGenre>>() {
        }.getType();
        List<TvShowGenresResponse.TvShowGenre> tvShowGenres = gson.fromJson(value, type);
        return tvShowGenres;
    }

    @TypeConverter
    public static String fromGenreForTvShowList(List<TvShowGenresResponse.TvShowGenre> tvShowGenres) {
        Gson gson = new Gson();
        String json = gson.toJson(tvShowGenres);
        return json;
    }

    @TypeConverter
    public static List<TvShowDetailsResponse.CreatedBy> toCreatedByList(String value) {
        if (value == null) {
            return (null);
        }
        Gson gson = new Gson();
        Type type = new TypeToken<List<TvShowDetailsResponse.CreatedBy>>() {
        }.getType();
        List<TvShowDetailsResponse.CreatedBy> createdByList = gson.fromJson(value, type);
        return createdByList;
    }

    @TypeConverter
    public static String fromCreatedByList(List<TvShowDetailsResponse.CreatedBy> createdByList) {
        Gson gson = new Gson();
        String json = gson.toJson(createdByList);
        return json;
    }
    @TypeConverter
    public static List<TvShowDetailsResponse.Season> toSeasonsList(String value) {
        if (value == null) {
            return (null);
        }
        Gson gson = new Gson();
        Type type = new TypeToken<List<TvShowDetailsResponse.Season>>() {
        }.getType();
        List<TvShowDetailsResponse.Season> seasonsList = gson.fromJson(value, type);
        return seasonsList;
    }

    @TypeConverter
    public static String fromSeasonsList(List<TvShowDetailsResponse.Season> seasonsList) {
        Gson gson = new Gson();
        String json = gson.toJson(seasonsList);
        return json;
    }

    @TypeConverter
    public static List<ActorSearchResponse.KnownFor> toKnownForList(String value) {
        if (value == null) {
            return (null);
        }
        Gson gson = new Gson();
        Type type = new TypeToken<List<ActorSearchResponse.KnownFor>>() {
        }.getType();
        List<ActorSearchResponse.KnownFor> knownForList = gson.fromJson(value, type);
        return knownForList;
    }

    @TypeConverter
    public static String fromKnownForList(List<ActorSearchResponse.KnownFor> knownForList) {
        Gson gson = new Gson();
        String json = gson.toJson(knownForList);
        return json;
    }
}

File: app/src/main/java/com/aero51/moviedatabase/utils/Converters.kt
package com.aero51.moviedatabase.utils

import androidx.room.TypeConverter
import com.aero51.moviedatabase.repository.model.omdb.OmdbModel.Ratings
import com.aero51.moviedatabase.repository.model.tmdb.credits.ActorSearchResponse.KnownFor
import com.aero51.moviedatabase.repository.model.tmdb.credits.MoviesWithPerson
import com.aero51.moviedatabase.repository.model.tmdb.credits.TvShowWithPerson
import com.aero51.moviedatabase.repository.model.tmdb.movie.MovieDetailsResponse
import com.aero51.moviedatabase.repository.model.tmdb.movie.MovieGenresResponse.MovieGenre
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowDetailsResponse
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowDetailsResponse.CreatedBy
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowDetailsResponse.Season
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowGenresResponse.TvShowGenre
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.util.*

//Room cannot store lists and Date among others  so type converters must be implemented
object Converters {
    @TypeConverter
    fun fromTimestamp(value: Long?): Date? {
        return if (value == null) null else Date(value)
    }

    @TypeConverter
    fun dateToTimestamp(date: Date?): Long? {
        return date?.time
    }

    @JvmStatic
    @TypeConverter
    fun fromString(value: String?): List<Int> {
        val listType = object : TypeToken<List<Int?>?>() {}.type
        return Gson().fromJson(value, listType)
    }

    @JvmStatic
    @TypeConverter
    fun fromArrayList(list: List<Int?>?): String {
        val gson = Gson()
        return gson.toJson(list)
    }

    @JvmStatic
    @TypeConverter
    fun fromString2(value: String?): List<String> {
        val listType = object : TypeToken<List<String?>?>() {}.type
        return Gson().fromJson(value, listType)
    }

    @JvmStatic
    @TypeConverter
    fun fromList(list: List<String?>?): String {
        val gson = Gson()
        return gson.toJson(list)
    }

    @JvmStatic
    @TypeConverter
    fun toRatingsList(value: String?): List<Ratings>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<Ratings?>?>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromRatingsList(ratings: List<Ratings?>?): String {
        val gson = Gson()
        return gson.toJson(ratings)
    }

    @JvmStatic
    @TypeConverter
    fun toMovieProductionCompaniesList(value: String?): List<MovieDetailsResponse.ProductionCompany>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<MovieDetailsResponse.ProductionCompany?>?>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromMovieProductionCompaniesList(productionCompanies: List<MovieDetailsResponse.ProductionCompany?>?): String {
        val gson = Gson()
        return gson.toJson(productionCompanies)
    }

    @JvmStatic
    @TypeConverter
    fun toTvShowProductionCompaniesList(value: String?): List<TvShowDetailsResponse.ProductionCompany>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<TvShowDetailsResponse.ProductionCompany?>?>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromTvShowProductionCompaniesList(productionCompanies: List<TvShowDetailsResponse.ProductionCompany?>?): String {
        val gson = Gson()
        return gson.toJson(productionCompanies)
    }

    @JvmStatic
    @TypeConverter
    fun toGenresForMovieList(value: String?): List<MovieGenre>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<MovieGenre?>?>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromGenreForMovieList(movieGenres: List<MovieGenre?>?): String {
        val gson = Gson()
        return gson.toJson(movieGenres)
    }

    @JvmStatic
    @TypeConverter
    fun toGenresForTvShowList(value: String?): List<TvShowGenre>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<TvShowGenre?>?>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromGenreForTvShowList(tvShowGenres: List<TvShowGenre?>?): String {
        val gson = Gson()
        return gson.toJson(tvShowGenres)
    }

    @JvmStatic
    @TypeConverter
    fun toCreatedByList(value: String?): List<CreatedBy>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<CreatedBy?>?>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromCreatedByList(createdByList: List<CreatedBy?>?): String {
        val gson = Gson()
        return gson.toJson(createdByList)
    }

    @JvmStatic
    @TypeConverter
    fun toSeasonsList(value: String?): List<Season>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<Season?>?>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromSeasonsList(seasonsList: List<Season?>?): String {
        val gson = Gson()
        return gson.toJson(seasonsList)
    }

    @JvmStatic
    @TypeConverter
    fun toKnownForList(value: String?): List<KnownFor>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<KnownFor?>?>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromKnownForList(knownForList: List<KnownFor?>?): String {
        val gson = Gson()
        return gson.toJson(knownForList)
    }

    @JvmStatic
    @TypeConverter
    fun toMoviesWithPersonCast(value: String?): List<MoviesWithPerson.Cast>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<MoviesWithPerson.Cast>>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromMoviesWithPersonCast(cast: List<MoviesWithPerson.Cast>): String {
        val gson = Gson()
        return gson.toJson(cast)
    }

    @JvmStatic
    @TypeConverter
    fun toMoviesWithPersonCrew(value: String?): List<MoviesWithPerson.Crew>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<MoviesWithPerson.Crew>>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromMoviesWithPersonCrew(crew: List<MoviesWithPerson.Crew>): String {
        val gson = Gson()
        return gson.toJson(crew)
    }
    @JvmStatic
    @TypeConverter
    fun toTvShowsWithPersonCrew(value: String?): List<TvShowWithPerson.Crew>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<TvShowWithPerson.Crew>>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromTvShowsWithPersonCrew(crew: List<TvShowWithPerson.Crew>): String {
        val gson = Gson()
        return gson.toJson(crew)
    }
    @JvmStatic
    @TypeConverter
    fun toTvShowsWithPersonCast(value: String?): List<TvShowWithPerson.Cast>? {
        if (value == null) {
            return null
        }
        val gson = Gson()
        val type = object : TypeToken<List<TvShowWithPerson.Cast>>() {}.type
        return gson.fromJson(value, type)
    }

    @JvmStatic
    @TypeConverter
    fun fromTvShowsWithPersonCast(cast: List<TvShowWithPerson.Cast>): String {
        val gson = Gson()
        return gson.toJson(cast)
    }
}